dcn-pytorch/utils.py

```diff
@@ -0,0 +1,211 @@
+from collections import OrderedDict, namedtuple, defaultdict
+
+
+import torch
+import torch.nn as nn
+from layers import SequencePoolingLayer
+
+
+DEFAULT_GROUP_NAME = "default_group"
+
+
+class SparseFeat(namedtuple('SparseFeat',
+                            ['name', 'vocabulary_size', 'embedding_dim', 'use_hash', 'dtype', 'embedding_name',
+                             'group_name'])):
+    __slots__ = ()
+
+    def __new__(cls, name, vocabulary_size, embedding_dim=4, use_hash=False, dtype="int32", embedding_name=None,
+                group_name=DEFAULT_GROUP_NAME):
+        if embedding_name is None:
+            embedding_name = name
+        if embedding_dim == "auto":
+            embedding_dim = 6 * int(pow(vocabulary_size, 0.25))
+        if use_hash:
+            print(
+                "Notice! Feature Hashing on the fly currently is not supported in torch version,you can use tensorflow version!")
+        return super(SparseFeat, cls).__new__(cls, name, vocabulary_size, embedding_dim, use_hash, dtype,
+                                              embedding_name, group_name)
+
+    def __hash__(self):
+        return self.name.__hash__()
```

Comment:
```suggestion
        return hash(self.name)
```

一般不推荐显式调用 `__xxx__` 这类特殊方法

Reply:
已修改